pkg/pma/pma.c

```diff
@@ -0,0 +1,783 @@
+///                    range must be within the bounds of the PMA.
+/// @param[in] status  New status of the page range in question. Must not be
+///                    PS_MASK.
+/// @param[in] pma     PMA the page range in question belongs to.
+static_ inline_ void
+set_page_status_range_(const void   *addr,
+                       size_t        pg_cnt,
+                       page_status_t status,
+                       const pma_t  *pma)
+{
+    for (size_t i = 0; i < pg_cnt * kPageSz; i += kPageSz) {
+        set_page_status_((char *)addr + i, status, pma);
+    }
+}
+
+/// Sync changes made to either the heap or stack since the last call to
+/// pma_sync().
+///
+/// @param[in] path        Path to backing file. If NULL, this function is a
+///                        no-op.
+/// @param[in] base        Base address of the file-backed mapping.
+/// @param[in] grows_down  true if the mapping grows downward in memory.
+/// @param[in] pma         PMA this mapping belongs to.
+/// @param[in] len         Number of bytes at the beginning of the mapping to be
+///                        synced to the backing file. Must be a multiple of
+///                        kPageSz.
+/// @param[in] fd          Open file descriptor of the backing file.
+///
+/// @return 0   Synced changes successfully.
+/// @return -1  Failed to sync changes.
+static_ int
+sync_file_(const char *path,
+           void       *base,
+           bool        grows_down,
+           pma_t      *pma,
+           size_t      len,
+           int         fd);
+
+/// Get the total length in bytes of a PMA.
+///
+/// @param[in] pma
+static_ inline_ size_t
+total_len_(const pma_t *pma)
+{
+    assert(pma);
+    return (size_t)(pma->stack_start - pma->heap_start);
+}
+
+static_ int
+center_guard_page_(pma_t *pma)
+{
+    size_t heap_len  = 0;
+    size_t stack_len = 0;
+    if (pma->len_getter(&heap_len, &stack_len) == -1) {
+        fprintf(stderr,
+                "pma: failed to determine length of heap and stack\r\n");
+        return -1;
+    }
+
+    size_t free_len   = total_len_(pma) - (heap_len + stack_len);
+    char  *free_start = (char *)pma->heap_start + pma->heap_len;
+    void  *guard_pg   = free_start + round_up(free_len / 2, kPageSz);
+
+    if (mmap(guard_pg,
+             kPageSz,
+             PROT_NONE,
+             MAP_ANON | MAP_FIXED | MAP_PRIVATE,
+             -1,
+             0)
+        == MAP_FAILED)
+    {
+        fprintf(stderr,
+                "pma: failed to place %zu-byte guard page at %p: %s\r\n",
+                kPageSz,
+                guard_pg,
+                strerror(errno));
+        return -1;
+    }
+    set_page_status_(guard_pg, PS_MAPPED_INACCESSIBLE, pma);
+    pma->guard_pg = guard_pg;
+    return 0;
+}
+
+static_ int
+handle_page_fault_(void *fault_addr, void *user_arg)
+{
+    fault_addr = (void *)round_down((uintptr_t)fault_addr, kPageSz);
+    pma_t *pma = user_arg;
+    switch (page_status_(fault_addr, pma)) {
```

Comment:
`clangd` warns about not handling the `PS_MAPPED_DIRTY` case here. Is it even possible for a page fault to happen on a dirty page?